### PR TITLE
New article to explain how to loadtest using OAuth.

### PR DESCRIPTION
**Pull request by Natim**
Could you guys please review my article?
@tarekziade @almet @leplatrem @n1k0 @rfk

### CODE REVIEW

File: content/2015.12.loadtesting.with.oauth.rst [hunk 1]
```diff
@@ -0,0 +1,214 @@
+Loadtesting a HTTP API which uses OAuth for authentication
```
Comment (almet) [inline content/2015.12.loadtesting.with.oauth.rst]: Consider: "How to load-test a OAuth-protected HTTP API?"

File: content/2015.12.loadtesting.with.oauth.rst [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+##########################################################
+
+:lang: en
+:date: 2015-12-02
+:url: loadtesting-a-http-api-which-uses-oauth-for-authentication
+:summary: How to loadtest an OAuth service relier?
+
+
+Why would you need to run loadtest?
```
Comment (almet) [inline content/2015.12.loadtesting.with.oauth.rst]: a loadtest

File: content/2015.12.loadtesting.with.oauth.rst [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+===================================
+
+After deploying an HTTP API service architecture you might be
```
Comment (almet) [inline content/2015.12.loadtesting.with.oauth.rst]: "an HTTP API" (removing "service architecture")

File: content/2015.12.loadtesting.with.oauth.rst [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+wondering how many users or requests your deploiement can take and
```
Comment (almet) [inline content/2015.12.loadtesting.with.oauth.rst]: deployment

File: content/2015.12.loadtesting.with.oauth.rst [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+what happens after it reaches that number.
+
+This is a great indicator to tell you to start scalling if you see
```
Comment (almet) [inline content/2015.12.loadtesting.with.oauth.rst]: nit: scaling

File: content/2015.12.loadtesting.with.oauth.rst [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+that your production server is close to reach that number of requests
+per seconds.
```
Comment (almet) [inline content/2015.12.loadtesting.with.oauth.rst]: consider: "Knowing the maximum load your service can handle is a great thing to know!"

File: content/2015.12.loadtesting.with.oauth.rst [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+
+
+Why is it a problem to do so with OAuth?
```
Comment (almet) [inline content/2015.12.loadtesting.with.oauth.rst]: Consider: "So, what's so complicated about OAuth?"

File: content/2015.12.loadtesting.with.oauth.rst [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+========================================
+
+Loadtesting with OAuth means having an OAuth account, often creating
```
Comment (almet) [inline content/2015.12.loadtesting.with.oauth.rst]: If you need to authenticate to the API using OAuth, then you'll need to create and validate an OAuth account. Then, you'll have access to Bearer Tokens.

Comment (almet) [inline content/2015.12.loadtesting.with.oauth.rst]: consider adding links to OAuth and Bearer Tokens, so the reader can understand what this is.

File: content/2015.12.loadtesting.with.oauth.rst [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+and validating it and then getting Bearer tokens with the right
+scopes.
+
+This flow will add a number of requests and waiting for the account
+activation email can take a tremendous amount of time in the scope of
+a loadtest which can slow down the all process and return inaccurate
+results.
```
Comment (almet) [inline content/2015.12.loadtesting.with.oauth.rst]: Consider rephrasing, this sentence is too long.

File: content/2015.12.loadtesting.with.oauth.rst [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+
+
+What is the right way to handle this?
+=====================================
+
+Depending on the tools you are using for your loadtest the specifics
```
Comment (almet) [inline content/2015.12.loadtesting.with.oauth.rst]: I would just drop this paragraph.

File: content/2015.12.loadtesting.with.oauth.rst [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+of the answer could be different but basically:
+
+During a loadtest, you should make sure that your are only doing
```
Comment (almet) [inline content/2015.12.loadtesting.with.oauth.rst]: s/you should//

File: content/2015.12.loadtesting.with.oauth.rst [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+requests on the service you are loadtesting.
+
+If you are calling an identity provider or any other service during
+the loadtest phases (either setup, tear down or scenario phase) your
+loadtest performance will be impacted and will return inaccurate
+results.
+
+Basically, you should be able to write your loadtest with simple cURL
```
Comment (almet) [inline content/2015.12.loadtesting.with.oauth.rst]: s/Basically, //

File: content/2015.12.loadtesting.with.oauth.rst [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+commands toward your service.
+
+
+Configuration over code
+=======================
+
+It means that if your service needs Bearer tokens to be loadtested,
```
Comment (almet) [inline content/2015.12.loadtesting.with.oauth.rst]: s/It means that//

File: content/2015.12.loadtesting.with.oauth.rst [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+environment variables.
+
+
+Some code to configure your loadtest
+====================================
+
+With regards to the kind of environment you want to loadtest or
+smoketest on (dev, stage, production), you may want to configure your loadtest differently.
+
+For instance, you may want to create specific account for your
```
Comment (almet) [inline content/2015.12.loadtesting.with.oauth.rst]: to create a specific account.

File: content/2015.12.loadtesting.with.oauth.rst [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+loadtest on stage and use an existing account for the production smoke
+test.
+
+
+Working with environment variables
+==================================
+
+With ailoads you can define a JSON file that will set the environment
```
Comment (almet) [inline content/2015.12.loadtesting.with.oauth.rst]: add a link to ailoads.

File: content/2015.12.loadtesting.with.oauth.rst [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+            raise
+
+	token = get_bearer_token(EMAIL, PASSWORD)
+    print('export OAUTH_BEARER_TOKEN="%s"' % token)
+
+
+Working with multiple Bearer Tokens
+===================================
+
+If you need to use multiple user, you can add other environment
```
Comment (almet) [inline content/2015.12.loadtesting.with.oauth.rst]: users

File: content/2015.12.loadtesting.with.oauth.rst [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+variable.
```
Comment (almet) [inline content/2015.12.loadtesting.with.oauth.rst]: variables

File: content/2015.12.loadtesting.with.oauth.rst [hunk 18]
```diff
@@ -0,0 +1,214 @@
...
+
+You can also set multiple users Bearer tokens separated by comma in
```
Comment (almet) [inline content/2015.12.loadtesting.with.oauth.rst]: multiple bearer tokens via using a comma-separated list in the env variable.

File: content/2015.12.loadtesting.with.oauth.rst [hunk 19]
```diff
@@ -0,0 +1,214 @@
...
+
+.. code-block:: bash
+
+    export OAUTH_BEARER_TOKEN="
+        dd7e6b6d0f4d44d0df5d6b73afdfbad41c48c4abdaba50e428ce070f9d4c75b5,
+        b6af04a44aa0f5a6b3a3affbaa41c48c4abaaba50e428ce030f9a4cb356aa36e,
+        a44aa0f5a6b3a3affbaa41c1c48c4abaaba50e428ce030f9428ce070f9d4c75b
+        "
+
+You can also create multiple environment variables:
```
Comment (almet) [inline content/2015.12.loadtesting.with.oauth.rst]: Or you also could

File: content/2015.12.loadtesting.with.oauth.rst [hunk 20]
```diff
@@ -0,0 +1,214 @@
...
+    export OAUTH_BEARER_TOKEN_SCOPE_PROFILE="dd7e6b6d0f4d44d0df5d6b73afdfbad41c48c4abdaba50e428ce070f9d4c75b5"
+    export OAUTH_BEARER_TOKEN_SCOPE_KINTO="
+        b6af04a44aa0f5a6b3a3affbaa41c48c4abaaba50e428ce030f9a4cb356aa36e,
+        a44aa0f5a6b3a3affbaa41c1c48c4abaaba50e428ce030f9428ce070f9d4c75b
+        "
+
+Conclusion
+==========
+
+After thinking this through, it doesn't look so complex to loadtest a
```
Comment (almet) [inline content/2015.12.loadtesting.with.oauth.rst]: I hope that after reading this article, you aren't afraid anymore of loadtesting an OAuth-based service!

Take-aways:

File: content/2015.12.loadtesting.with.oauth.rst [hunk 21]
```diff
@@ -0,0 +1,214 @@
+Loadtesting a HTTP API which uses OAuth for authentication
+##########################################################
+
+:lang: en
+:date: 2015-12-02
+:url: loadtesting-a-http-api-which-uses-oauth-for-authentication
+:summary: How to loadtest an OAuth service relier?
```
Comment (leplatrem) [inline content/2015.12.loadtesting.with.oauth.rst]: feedback: a while ago, I would have no idea what a relier means :) Since the summary appears in the homepage, you can be a bit more talkative like «This article explains how to test the performance of an HTTP API when its access depends on an external OAuth service»

File: content/2015.12.loadtesting.with.oauth.rst [hunk 22]
```diff
@@ -0,0 +1,214 @@
...
+Why would you need to run a loadtest?
+=====================================
+
+After deploying an HTTP API you might be wondering how many users or
+requests your deployment can take and what happens after it reaches
+that number.
+
+Knowing the maximum load your service can handle is a great thing to
+know!  It tells you when to start scaling up when your production
+server is close to reach that number of requests per seconds.
```
Comment (tarekziade) [inline content/2015.12.loadtesting.with.oauth.rst]: per second

File: content/2015.12.loadtesting.with.oauth.rst [hunk 23]
```diff
@@ -0,0 +1,214 @@
+How to load-test a OAuth-protected HTTP API?
+############################################
+
+:lang: en
+:date: 2015-12-02
+:url: loadtesting-a-http-api-which-uses-oauth-for-authentication
+:summary: This article explains how to test the performance of an HTTP API when its access depends on an external OAuth service.
+
+
```
Comment (tarekziade) [inline content/2015.12.loadtesting.with.oauth.rst:10]: I feel like we miss a section here that briefly explains what is OAuth and how it's used to authenticate to an HTTP API

File: content/2015.12.loadtesting.with.oauth.rst [hunk 24]
```diff
@@ -0,0 +1,214 @@
...
+Knowing the maximum load your service can handle is a great thing to
+know!  It tells you when to start scaling up when your production
+server is close to reach that number of requests per seconds.
+
+
+So, what's so complicated about Oauth?
+======================================
+
+Loadtesting with `OAuth <https://en.wikipedia.org/wiki/OAuth>`_ means
```
Comment (tarekziade) [inline content/2015.12.loadtesting.with.oauth.rst]: s/Loadtesting with/ Loadtesting a service that requires OAuth-based authentication/

File: content/2015.12.loadtesting.with.oauth.rst [hunk 25]
```diff
@@ -0,0 +1,214 @@
...
+having an OAuth account, often creating and validating it and then
```
Comment (tarekziade) [inline content/2015.12.loadtesting.with.oauth.rst]: having xxx, often xx and xx and then xxx . That sentence is hard to follow. I would split it or use bullet points

File: content/2015.12.loadtesting.with.oauth.rst [hunk 26]
```diff
@@ -0,0 +1,214 @@
...
+getting `Bearer tokens <http://tools.ietf.org/html/rfc6750>`_ with the
+right scopes.
+
+Doing all this setup, will add a number of requests. Plus it will take
+time to wait for the email with the activation account link.
+
+This can take a tremendous amount of time in the scope of a loadtest
```
Comment (tarekziade) [inline content/2015.12.loadtesting.with.oauth.rst]: s/loadtest/load test/  (and in other places in the doc)

File: content/2015.12.loadtesting.with.oauth.rst [hunk 27]
```diff
@@ -0,0 +1,214 @@
...
+
+What is the right way to handle this?
+=====================================
+
+During a loadtest, make sure that your are only doing requests on the
+service you are loadtesting.
+
+If you are calling an identity provider or any other service during
+the loadtest phases (either setup, tear down or scenario phase) your
+loadtest performance will be impacted and will return inaccurate
```
Comment (tarekziade) [inline content/2015.12.loadtesting.with.oauth.rst]: maybe we should explain here that we want to measure the API performance in isolation, not the full flow. Because we could load test the whole flow, including calls to a 3rd party

File: content/2015.12.loadtesting.with.oauth.rst [hunk 28]
```diff
@@ -0,0 +1,214 @@
...
+
+You should be able to write your loadtest with simple cURL commands
+toward your service.
+
+
+Configuration over code
+=======================
+
+If your service needs Bearer tokens to be loadtested, these Bearer
+tokens should be provided to your code loadtest and not be generated
```
Comment (tarekziade) [inline content/2015.12.loadtesting.with.oauth.rst]: s/code loadtest/load test code/

File: content/2015.12.loadtesting.with.oauth.rst [hunk 29]
```diff
@@ -0,0 +1,214 @@
...
+
+This can be done by configuration, either within a file or with
+environment variables.
+
+
+Some code to configure your loadtest
+====================================
+
+With regards to the kind of environment you want to loadtest or
+smoketest on (dev, stage, production), you may want to configure your
```
Comment (tarekziade) [inline content/2015.12.loadtesting.with.oauth.rst]: smoke test is introduced here but was not explained. It's vague what's the difference with load tests in this section

File: content/2015.12.loadtesting.with.oauth.rst [hunk 30]
```diff
@@ -0,0 +1,214 @@
...
+loadtest differently.
+
+For instance, you may want to create specific account or use an
```
Comment (tarekziade) [inline content/2015.12.loadtesting.with.oauth.rst]: a/specific account/a specific account /

File: content/2015.12.loadtesting.with.oauth.rst [hunk 31]
```diff
@@ -0,0 +1,214 @@
...
+existing one.
+
+
+Working with environment variables
+==================================
+
+With `ailoads <https://github.com/tarekziade/ailoads>`_ you can define
```
Comment (tarekziade) [inline content/2015.12.loadtesting.with.oauth.rst]: you should introduce aioloads in one sentence. It's coming out of the blue here

File: content/2015.12.loadtesting.with.oauth.rst [hunk 32]
```diff
@@ -0,0 +1,214 @@
...
+a JSON file that will set the environment variables before running the
+docker instances.
+
+Also when you want to run it without using loadsv2 it can be a problem
```
Comment (tarekziade) [inline content/2015.12.loadtesting.with.oauth.rst]: same thing for loadsv2

**Comment by tarekziade** [comment]
My general feeling about this article is that it has a lot of very good stuff and tips, but it's hard to follow because the narrative is half-way between a tutorial and a "bag of tricks"

Maybe it could be a step-by-step tutorial where you explain how you can load test an Oauth-based HTTP Service using ailoads and loadvs2, with a section for each step? and a intro presenting your use case and the tools you are going to use through the tutorial

My 2 cents

File: content/2015.12.loadtesting.with.oauth.rst (hunk 23 above)
Comment (Natim) [inline content/2015.12.loadtesting.with.oauth.rst:10]: :+1:

File: content/2015.12.loadtesting.with.oauth.rst (hunk 27 above)
Comment (Natim) [inline content/2015.12.loadtesting.with.oauth.rst]: Actually the load-test will still include calls to 3rd party on server side but not on client side.

**Comment by Natim** [comment]
> Maybe it could be a step-by-step tutorial where you explain how you can load test an Oauth-based HTTP Service using ailoads and loadvs2, with a section for each step?

I wonder if this shouldn't be the scope of another article after I have actually build the tooling around loadsv2?

I removed the tutorial part from the article. What do you think?

File: content/2015.12.loadtesting.with.oauth.rst [hunk 33]
```diff
@@ -0,0 +1,171 @@
...
+:lang: en
+:date: 2015-12-02
+:url: load-testing-a-http-api-which-uses-oauth-for-authentication
+:summary: This article explains how to test the performance of an HTTP API when its access depends on an external OAuth service.
+
+
+What is OAuth and how do you use it to authenticate an HTTP API?
+================================================================
+
+`OAuth <https://en.wikipedia.org/wiki/OAuth>`_ is a protocol that let
```
Comment (rfk) [inline content/2015.12.loadtesting.with.oauth.rst]: s/let/lets/, but you might also consider re-wording to something like "allows a user to delegate"

File: content/2015.12.loadtesting.with.oauth.rst [hunk 34]
```diff
@@ -0,0 +1,171 @@
...
+a user delegate some of its permissions to a web service through an
+identity provider.
+
+The user can be confident that the web service will never have its
```
Comment (rfk) [inline content/2015.12.loadtesting.with.oauth.rst]: s/its/their/

File: content/2015.12.loadtesting.with.oauth.rst [hunk 35]
```diff
@@ -0,0 +1,171 @@
...
+Why would you need to run a load test?
+======================================
+
+After deploying an HTTP API you might be wondering how many users or
+requests your deployment can take and what happens after it reaches
+that number.
+
+Knowing the maximum load your service can handle is a great thing to
+know!  It tells you when to start scaling up when your production
+server is close to reach that number of requests per second.
```
Comment (rfk) [inline content/2015.12.loadtesting.with.oauth.rst]: s/reach/reaching/

File: content/2015.12.loadtesting.with.oauth.rst [hunk 36]
```diff
@@ -0,0 +1,171 @@
...
+
+
+So, what's so complicated about OAuth?
+======================================
+
+Load testing a service that requires OAuth-based authentication means:
+
+* having an OAuth account (often creating it) 
+* getting a Bearer Token with the right scopes
```
Comment (rfk) [inline content/2015.12.loadtesting.with.oauth.rst:62]: A related concern (which I don't know whether you're dealing with or not in your use-case) is whether you test load with a single user account, or with lots of different user accounts.  The difference can affect the performance of your app, and using many different accounts is harder but more realistic.

File: content/2015.12.loadtesting.with.oauth.rst [hunk 37]
```diff
@@ -0,0 +1,171 @@
...
+.. code-block:: bash
+
+    docker run -e OAUTH_BEARER_TOKEN="abc..." loadtest
+
+
+Working with multiple Bearer Tokens
+===================================
+
+If you need to use multiple users, you can add other environment
+variables.
```
Comment (rfk) [inline content/2015.12.loadtesting.with.oauth.rst:138]: heh, I see, you talk about multiple users a little down here :-)

**Comment by rfk** [comment]
This cut-down version presents some good things to consider when loadtesting an oauth service, but it doesn't seem to be properly "framed".  The title promises "How to load test an OAuth-protected HTTP API", but the content is more about "things to keep in mind" than specific "how-tos".

If you're planning a follow-up article with more details around the specific tools etc you're using, I think it would be good to say that explicitly, both at the start with a "this is part 1 of a series on loadtesting oauth APIs" and at the end with a "the next article will show some specific tools you can use to do this".
